### PR TITLE
cram-md5: do not accept challenge if own hostname is used

### PR DESCRIPTION
**Pull request by franku**
based on master

### CODE REVIEW

File: core/src/lib/cram_md5.cc [hunk 1]
```diff
@@ -39,10 +40,34 @@ CramMd5Handshake::CramMd5Handshake(BareosSocket* bs,
...
+CramMd5Handshake::ComparisonResult
+CramMd5Handshake::CompareChallengeWithOwnQualifiedName(
+    const char* challenge) const
+{
+  uint32_t a, b;
+  char buffer[128]{"?"};  // at least one character
+
+  bool scan_success = sscanf(challenge, "<%u.%u@%s", &a, &b, buffer) == 3;
+
+  // string contains the closing ">" of the challange
```
Comment (pstorz) [inline core/src/lib/cram_md5.cc:57]: challenge

File: core/src/lib/cram_md5.cc [hunk 2]
```diff
@@ -39,10 +40,34 @@ CramMd5Handshake::CramMd5Handshake(BareosSocket* bs,
...
+  std::string challenge_qualified_name(buffer, strlen(buffer) - 1);
+
+  Dmsg1(debuglevel_, "my_name: <%s> - challenge_name: <%s>",
+        own_qualified_name_bashed_spaces_.c_str(),
+        challenge_qualified_name.c_str());
+
+  if (!scan_success) { return ComparisonResult::FAILURE; }
+
+  // check if the name in the challange matches the daemons name
```
Comment (pstorz) [inline core/src/lib/cram_md5.cc:66]: challenge

File: core/src/lib/cram_md5.cc [hunk 3]
```diff
@@ -167,6 +190,15 @@ bool CramMd5Handshake::CramMd5Response()
     }
   }
 
+  auto result = CompareChallengeWithOwnQualifiedName(chal.c_str());
+
+  if (result == ComparisonResult::IS_SAME ||
```
Comment (pstorz) [inline core/src/lib/cram_md5.cc:195]: I would suggest to log a warning when IS_SAME ist true like "Warning: detected attack:  I will not answer to my own challenge. Please check integrity of <name of client that wants to connnect>"

File: core/src/lib/bsock.cc [hunk 4]
```diff
@@ -404,8 +404,37 @@ bool BareosSocket::TwoWayAuthenticate(JobControlRecord* jcr,
...
-      Jmsg(jcr, M_FATAL, 0, _("Authorization key rejected %s.\n"), identity);
+      char ipaddr_str[48]{};
+      SockaddrToAscii(&(client_addr), ipaddr_str, sizeof(ipaddr_str));
+
+      switch (cram_md5_handshake.result) {
+        case CramMd5Handshake::HandshakeResult::CHALLENGE_ATTACK: {
+          const char* fmt =
+              "Warning! Attack detected: "
+              "I will not answer to my own challenge. "
+              "Please check integrity of IP Address: %s\n";
```
Comment (franku) [inline core/src/lib/bsock.cc:425]:
```suggestion
              "Please check integrity of the host at IP address: %s\n";
```

File: core/src/lib/bsock.cc [hunk 5]
```diff
@@ -404,8 +404,37 @@ bool BareosSocket::TwoWayAuthenticate(JobControlRecord* jcr,
     }
 
     auth_success = cram_md5_handshake.DoHandshake(initiated_by_remote);
+
     if (!auth_success) {
-      Jmsg(jcr, M_FATAL, 0, _("Authorization key rejected %s.\n"), identity);
+      char ipaddr_str[48]{};
```
Comment (franku) [inline core/src/lib/bsock.cc:417]:
```suggestion
      char ipaddr_str[MAXHOSTNAMELEN]{};
```

File: core/src/lib/cram_md5.cc [hunk 6]
```diff
@@ -39,10 +40,34 @@ CramMd5Handshake::CramMd5Handshake(BareosSocket* bs,
...
-  return;
+  BashSpaces(own_qualified_name_bashed_spaces_);
 }
 
+CramMd5Handshake::ComparisonResult
+CramMd5Handshake::CompareChallengeWithOwnQualifiedName(
+    const char* challenge) const
+{
+  uint32_t a, b;
+  char buffer[128]{"?"};  // at least one character
```
Comment (franku) [inline core/src/lib/cram_md5.cc:53]:
```suggestion
  char buffer[MAXHOSTNAMELEN]{"?"};  // at least one character
```

**Comment by franku** [comment]
Force pushed after hands-on review with arogge.

**Pull request merged**